File: day03/challenge1.py
#!/usr/bin/env python3

import sys
import re

def main():
    #read input from stdin
    lines = [line.strip() for line in sys.stdin]

    #a regular expression that captures every instance of mul() with two 1-3 digit numbers as arguments
    valid_exp = r"(mul\(\d{1,3}\s*,\s*\d{1,3}\))"

    #a regular expression that extracts the two numeric arguments to mul()
    mul_exp = r"mul\(\s*(\d{1,3})\s*,\s*(\d{1,3})"

    sum = 0
    #find all instances of the above regex in the input
    for line in lines:
        matches = [match.groups() for match in re.finditer(valid_exp, line)]
        print(matches)
        for match in matches:
            #extract the arguments to mul from match
            mul_args = re.search(mul_exp, match[0]).groups()
        
            print(f'{mul_args}')
            sum += int(mul_args[0])*int(mul_args[1]) 


    print(sum)

if __name__ == "__main__":
    main()

File: day03/challenge1.go
package main

import (
	"fmt"
	"regexp"
	"strconv"
)

func main() {
	//read input from stdin into an array of strings
	var lines []string
	for {
		var line string
		fmt.Scan(&line)
		if len(line) == 0 {
			break
		}
		lines = append(lines, line)
	}
	//regular expression that matches mul(\d{1,3},\d{1,3})
	var re = regexp.MustCompile(`mul\(([0-9]+),([0-9]+)\)`)
	sum := int64(0)
	//iterate over each line in the = 0
	for _, line := range lines {
		//match each occurance of the regular expression against the line
		for _, match := range re.FindAllStringSubmatch(line, -1) {
			//convert matched numbers to integers and multiply them
			a, err := strconv.ParseInt(match[1], 10, 64)
			if err != nil {
				fmt.Println("Error parsing a: ", err)
				return
			}
			b, err := strconv.ParseInt(match[2], 10, 64)
			if err != nil {
				fmt.Println("Error parsing b: ", err)
				return
			}
			sum += a * b
		}
	}
	fmt.Println("Sum:  ", sum)
}
